include/swift/SIL/SILInstruction.h

```diff
@@ -693,6 +745,184 @@ inline SingleValueInstruction *SILNode::castToSingleValueInstruction() {
+           kind <= SILNodeKind::Last_MultipleValueInstruction;
+  }
+};
+
+/// A utility mixin class that must be used by /all/ subclasses of
+/// MultipleValueInstruction to store their results.
+template <SILInstructionKind Kind, typename Derived, typename DerivedResult,
+          typename... OtherTrailingTypes>
+class MultipleValueInstructionTrailingObjects
+    : protected llvm::TrailingObjects<Derived, MultipleValueInstruction *,
+                                      DerivedResult, OtherTrailingTypes...> {
+  static_assert(LLVM_IS_FINAL(DerivedResult),
+                "Expected DerivedResult to be final");
+  static_assert(
+      std::is_base_of<MultipleValueInstructionResult, DerivedResult>::value,
+      "Expected DerivedResult to be a subclass of "
+      "MultipleValueInstructionResult");
+  static_assert(sizeof(MultipleValueInstructionResult) == sizeof(DerivedResult),
+                "Expected DerivedResult to be the same size as a "
+                "MultipleValueInstructionResult");
+
+protected:
+  using TrailingObjects =
+      llvm::TrailingObjects<Derived, MultipleValueInstruction *, DerivedResult,
+                            OtherTrailingTypes...>;
+  friend TrailingObjects;
+
+  using TrailingObjects::totalSizeToAlloc;
+  using TrailingObjects::getTrailingObjects;
+
+  unsigned NumResults;
+
+  size_t numTrailingObjects(typename TrailingObjects::template OverloadToken<
+                            MultipleValueInstruction *>) const {
+    return 1;
+  }
+
+  size_t numTrailingObjects(
+      typename TrailingObjects::template OverloadToken<DerivedResult>) const {
+    return NumResults;
+  }
+
+  template <typename... Args>
+  MultipleValueInstructionTrailingObjects(
+      Derived *Parent, ArrayRef<SILType> Types,
+      ArrayRef<ValueOwnershipKind> OwnershipKinds, Args &&... OtherArgs)
+      : NumResults(Types.size()) {
+
+    // If we do not have any results, then we do not need to initialize even the
+    // parent pointer since we do not have any results that will attempt to get
+    // our parent pointer.
+    if (!NumResults)
+      return;
+
+    auto **ParentPtr =
+        this->template getTrailingObjects<MultipleValueInstruction *>();
+    *ParentPtr = static_cast<MultipleValueInstruction *>(Parent);
+
+    auto *DataPtr = this->template getTrailingObjects<DerivedResult>();
+    for (unsigned i : range(NumResults)) {
+      ::new (&DataPtr[i]) DerivedResult(i, Types[i], OwnershipKinds[i],
+                                        std::forward<Args>(OtherArgs)...);
+      assert(DataPtr[i].getParent() == Parent &&
+             "Failed to setup parent reference correctly?!");
+    }
+  }
+
+  // Destruct the Derived Results.
+  ~MultipleValueInstructionTrailingObjects() {
+    if (!NumResults)
+      return;
+    auto *DataPtr = this->template getTrailingObjects<DerivedResult>();
+    for (unsigned i : range(NumResults))
+      DataPtr[i].~DerivedResult();
```

Comment:
Probably worth leaving a comment that this is just for the assertion.

Reply:
There is no assertion here. Can you elaborate?

Reply:
I did the rest of the changes here: https://github.com/apple/swift/pull/12612. Once you elaborate, I'll take care of this.

Reply:
The assertion in ``~ValueBase()``.